### PR TITLE
Register terminal panel before restoring serialized terminals

### LINKED ISSUE #62435: Zed won't open a terminal pane
### Reproduction steps

- Start zed
- Press Command + R 
- Observe a Git pane opening
- Cmd + P to open command prompt and press `terminal panel: toggle`
- observe nothing happening
- Cmd + P `workspace: new terminal`
- observe nothing happening

### Current vs. Expected behavior

- I expect the terminal pane to open when pressing `cmd + R`
- I expect the terminal pane to toggle when toggled
- I expect a new terminal to open when created

**NOTE** about 3/4 of the way through of writing this issue the correct behavior occurred. 

I think it is possible that the same hangs in https://github.com/zed-industries/zed/issues/55211 are preventing the terminal pane from opening here.

### Zed version and system specs

Zed Preview 1.15.0 
Commit: 05d125a865cbfd8a498278f8e396b22ea554ccaa
Version: 1.15.0+preview.336.05d125a865cbfd8a498278f8e396b22ea554ccaa

### Attach Zed log file

[zed.log](https://github.com/user-attachments/files/30903392/zed.log)

### Relevant Zed settings

<details><summary>settings.json</summary>

```json
// Zed settings
//
// For information on how to configure Zed, see the Zed
// documentation: https://zed.dev/docs/configuring-zed
//
// To see all of Zed's default settings without changing your
// custom settings, run `zed: open default settings` from the
// command palette (cmd-shift-p / ctrl-shift-p)
{
  "cli_default_open_behavior": "new_window",
  "terminal": {
    "dock": "right"
  },
  "ssh_connections": [
    {
      "host": "jackalope",
      "username": "josi",
      "args": [],
      "projects": [
        {
          "paths": [
            "/home/josi/git/ricochet/"
          ]
        }
      ],
      "port_forwards": [
        { "local_port": 6188, "remote_port": 3334 }
      ]
    }
  ],
  "project_panel": {
    "dock": "left"
  },
  "outline_panel": {
    "dock": "left"
  },
  "icon_theme": {
    "mode": "dark",
    "light": "Catppuccin Mocha",
    "dark": "Catppuccin Mocha"
  },
  "session": {
    "trust_all_worktrees": true
  },
  "ui_font_size": 14.0,
  "buffer_font_size": 12.0,
  "theme": {
    "mode": "dark",
    "light": "One Light",
    "dark": "Catppuccin Mocha",
  },
}

```

</details>

### Relevant Keymap

<details><summary>keymap.json</summary>

```json

```

</details>

### (for AI issues) Model provider details

_No response_

### If you are using WSL on Windows, what flavor of Linux are you using?

None

### PR DESCRIPTION
**Pull request by SomeoneToIgnore**
Closes https://github.com/zed-industries/zed/issues/62435
Part of https://github.com/zed-industries/zed/issues/60548

Before:

https://github.com/user-attachments/assets/6b5147da-44cb-40c7-9da5-c50145e2ab4e

After:

https://github.com/user-attachments/assets/c58423cc-ea99-40d9-ad18-23b31dfb8e5d

Previously, `TerminalPanel::load` awaited full deserialization of serialized
terminals (each awaiting directory environment capture via a login shell)
before the panel was registered in the dock. 

Until then, `terminal_panel::Toggle{Focus}` was a silent no-op: the workspace finds no
panel in any dock, the keystroke is eaten, and nothing happens.

When it got eventually restored, it would close whatever panel in the dock was open manually instead, wrecking the workflow.

Repro:
1. Add `case "$ZSH_EXECUTION_STRING" in *--printenv*) sleep 10;; esac` to `~/.zshrc`.
2. Open a project, open at least one terminal in the panel, quit Zed (Cmd+Q).
3. Relaunch via Dock/Spotlight (not the `zed` CLI, which skips env capture via
   `cli_environment`).
4. Press the terminal shortcut within 10s: nothing happens; after ~10s the
   panel appears and the shortcut works.

* `Register terminal panel before restoring serialized terminals` — the panel
  is now created and registered immediately, and restoration runs in
  panel-owned tasks (cancelled on panel drop, nothing detached).
  Terminal-adding paths were initially queued behind restoration (later
  replaced by grafting, see below); the default shell spawns only if the
  panel ends up active and empty after restore.

* `Stop replaying serialized dock state over manual panel changes` — fixes a
  related race: `Dock::add_panel` replayed the serialized dock state (active
  panel, visibility, zoom) on every panel insertion, so a late-loading panel
  would re-activate the serialized panel over one the user had switched to in
  the meantime (e.g. the terminal panel yanking the dock away from the agent
  panel), or re-open a dock the user had closed. Serialized state is now
  applied at most once and is discarded as soon as the user changes the dock
  state manually. The new tests fail without this change and pass with it.

* `Show restoration progress in the terminal panel and allow spawning
  terminals during it` — while terminals restore, the panel shows a spinner
  with "Restoring terminals…" instead of a blank pane, and logs the restored
  count and elapsed time. Supersedes the queuing from the first commit: new
  terminals spawn immediately, and any that land mid-restore are grafted into
  the restored layout instead of being lost. Serialization is paused during
  restoration, so quitting mid-restore can no longer persist a partial tab
  list and silently drop terminals.

Release Notes:

- Fixed the terminal panel not opening at all under certain circumstances, made panels no longer switching back to the serialized one when opened during startup.

### CODE REVIEW

File: crates/terminal_view/src/terminal_panel.rs [hunk 1]
```diff
@@ -254,76 +326,96 @@ impl TerminalPanel {
...
-        };
-
-        if let Some(workspace) = workspace.upgrade() {
-            workspace.update(&mut cx, |workspace, _| {
-                workspace.set_terminal_provider(TerminalProvider(terminal_panel.clone()))
-            });
+                .await;
+            if let Some(restored_terminals) = deserialized.log_err() {
+                restored = restored_terminals > 0;
+                log::info!(
```
Comment (MrSubidubi) [inline crates/terminal_view/src/terminal_panel.rs]: Do we want this to be an info-log?

File: crates/terminal_view/src/terminal_panel.rs [hunk 2]
```diff
@@ -793,52 +885,58 @@ impl TerminalPanel {
...
                     }
-                    RevealStrategy::Never => {}
-                }
 
-                pane.update(cx, |pane, cx| {
-                    let focus = matches!(reveal_strategy, RevealStrategy::Always);
-                    pane.add_item(terminal_view, true, focus, None, window, cx);
-                });
+                    pane.update(cx, |pane, cx| {
+                        let focus = matches!(reveal_strategy, RevealStrategy::Always);
```
Comment (MrSubidubi) [inline crates/terminal_view/src/terminal_panel.rs]: Should use Eq here

File: crates/terminal_view/src/terminal_panel.rs [hunk 3]
```diff
@@ -1286,6 +1390,49 @@ async fn wait_for_terminals_tasks(
     join_all(pending_tasks).await;
 }
 
+struct PendingTerminalGuard {
+    terminal_panel: Option<WeakEntity<TerminalPanel>>,
+    cx: AsyncApp,
+}
```
Comment (MrSubidubi) [inline crates/terminal_view/src/terminal_panel.rs]: Okay, this somewhat reminds me of a lecture about monads (or `with` in Python), so let's talk about it - from what I see this attempts is basically guard an operation - why does it not own said async operation, execute it, then handle the Ok and err-case appropriately? Not insisting, but this seems easy to misuse and also error-prone of sorts, also big parts above are just sheer duplication. Its only n = 2, but that still seems odd.

Also, why an option here? Is that to prevent racing with the drop impl? That seems so odd

File: crates/terminal_view/src/terminal_panel.rs [hunk 4]
```diff
@@ -1372,11 +1519,42 @@ impl Render for TerminalPanel {
...
+                    Icon::new(IconName::ArrowCircle)
+                        .color(Color::Muted)
+                        .size(IconSize::Small)
+                        .with_rotate_animation(2),
+                )
+                .child(Label::new(label).color(Color::Muted))
+                .with_animation(
+                    "restoring-terminals-fade-in",
+                    Animation::new(Duration::from_millis(300)),
+                    |this, delta| this.opacity(delta),
```
Comment (MrSubidubi) [inline crates/terminal_view/src/terminal_panel.rs]: We rotate and also dim both text and the rotated element?

File: crates/workspace/src/dock.rs [hunk 5]
```diff
@@ -437,6 +439,8 @@ impl Dock {
                 focus_follows_mouse: WorkspaceSettings::get_global(cx).focus_follows_mouse,
                 _subscriptions: [focus_subscription, zoom_subscription],
                 serialized_dock: None,
+                restoring_state: false,
+                restoration_finished: false,
```
Comment (MrSubidubi) [inline crates/workspace/src/dock.rs]: Not like this is not equally easy to misuse, but here at least wondering whether we could not combine this state better, e.g., does this not allow modelling state that should not ever be possible?

**Review by MrSubidubi — CHANGES_REQUESTED** [review]
Image

File: crates/terminal_view/src/terminal_panel.rs (hunk 3 above)
Comment (SomeoneToIgnore) [reply crates/terminal_view/src/terminal_panel.rs]: I thin this was the worst part (came in the tidy up follow-up), removed entirely and added `TerminalPanel::spawn_pending_terminal` instead.

File: crates/terminal_view/src/terminal_panel.rs (hunk 4 above)
Comment (SomeoneToIgnore) [reply crates/terminal_view/src/terminal_panel.rs]: Not anymore.

File: crates/workspace/src/dock.rs (hunk 5 above)
Comment (SomeoneToIgnore) [reply crates/workspace/src/dock.rs]: Reworked into `enum DockRestoreState`.

File: crates/terminal_view/src/terminal_panel.rs [hunk 6]
```diff
@@ -969,73 +955,100 @@ impl TerminalPanel {
...
+                cx.spawn(async move |cx| {
+                    terminal_panel
+                        .update(cx, |terminal_panel, cx| {
+                            terminal_panel.finish_pending_terminal(cx)
+                        })
+                        .ok();
+                })
+                .detach();
+            }
+        });
```
Comment (MrSubidubi) [inline crates/terminal_view/src/terminal_panel.rs:1026]: Checking into this, I am not sure we need the logic at all - we seemt to currently only be detaching that task and logging an error if it errors, nothing else. If we end up dropping the task, we have also dropped the app, so this will fail anyway, right?

Curious whether we can perhaps even clean up the call sites for `add_terminal_shell_internal` because both usages `add_terminal_shell` and `add_local_terminal_shell` are also just invoked at the same places twice, so could perhaps clean that up too. 

It might make sense to remove the possibility through a wrapper to not detatch the task as to prevent this drop logic from becoming relevant, but curious about what **you** think.

**Review by MrSubidubi — CHANGES_REQUESTED** [review]
Like the state on the workspace much better, yet some other thought to the other place.

**Review by MrSubidubi — APPROVED** [review]
